File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/core/ExposureStateImpl.kt
package dev.zeekr.camerax_android.core

import androidx.camera.core.ExposureState
import dev.zeekr.camerax_android.CameraXApiPigeonProxyApiRegistrar
import dev.zeekr.camerax_android.PigeonApiExposureStateApi
import dev.zeekr.camerax_android.common.IntRange

class ExposureStateImpl(registrar: CameraXApiPigeonProxyApiRegistrar) : PigeonApiExposureStateApi(registrar) {
    override fun exposureCompensationIndex(pigeon_instance: ExposureState): Long {
        return pigeon_instance.exposureCompensationIndex.toLong()
    }

    override fun exposureCompensationRange(pigeon_instance: ExposureState): IntRange {
        return IntRange(pigeon_instance.exposureCompensationRange)
    }

    override fun exposureCompensationStep(pigeon_instance: ExposureState): Double {
        return pigeon_instance.exposureCompensationStep.toDouble()
    }

    override fun isExposureCompensationSupported(pigeon_instance: ExposureState): Boolean {
        return pigeon_instance.isExposureCompensationSupported
    }
}

File: camerax_android/lib/src/core/exposure_state_impl.dart
import 'package:camerax_android/src/camerax_api.g.dart';
import 'package:camerax_android/src/common.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class ExposureStateImpl extends ExposureState {
  final ExposureStateProxyApi api;

  ExposureStateImpl.internal(this.api) : super.impl();

  @override
  int get exposureCompensationIndex => api.exposureCompensationIndex;
  @override
  RangeApi get exposureCompensationRange => api.exposureCompensationRange.impl;
  @override
  double get exposureCompensationStep => api.exposureCompensationStep;
  @override
  bool get isExposureCompensationSupported =>
      api.isExposureCompensationSupported;
}

extension ExposureStateProxyApiX on ExposureStateProxyApi {
  ExposureState get impl => ExposureStateImpl.internal(this);
}
